File: backend/app/cakisma.py
"""Aynı anda olma ölçütü: iki yerleşim ne zaman çakışır?

Bir şube ya da bir öğretmen aynı anda iki yerde olamaz. "Aynı an"ın ne demek
olduğu ise okula göre değişir; kurum dönem ayarından seçer
(bkz. `Term.conflict_basis`):

  * ``ders_saati`` — ızgaranın satırı esastır. Salı 3. ders yalnızca Salı
    3. dersle çakışır; saat bilgisi hiç okunmaz. Varsayılan ve eski davranış.
    Tek bir düzenli ızgarası olan okullar için doğru olan budur: satır zaten
    saatin kendisidir.
  * ``saat`` — gerçek saat aralığı esastır. Aynı gün içinde 09:00–09:40 ile
    09:20–10:00 çakışır, ayrı satırlar olsalar bile.

İkincisi, saatleri üst üste binebilen ızgaralarda anlamlıdır: vardiyalı okullar,
bir bölümü 40 diğer bölümü 30 dakikalık dersler, ya da elle girilirken kayan
saatler. Saati girilmemiş satır ``saat`` ölçütünde de yalnızca kendisiyle
çakışır — bilinmeyen bir aralık hakkında bir şey söylenemez, uydurmak yerine
satır kimliğine düşülür.

Ölçüt tek yerde tanımlıdır; hem çözücü (`app.solver.engine`) hem elle düzenleme
(`app.duzenle`) buradan okur. Yoksa ikisi farklı şeye "çakışma" derdi.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import time
from typing import Sequence

# Dönem ayarının alabileceği değerler.
DERS_SAATI = "ders_saati"
SAAT = "saat"


@dataclass(frozen=True)
class Aralik:
    """Bir ders saatinin zaman üzerindeki yeri.

    `baslangic`/`bitis` gün başından beri dakikadır; ikisi de yoksa satırın
    saati tanımsızdır.
    """
    gun_index: int
    baslangic: int | None = None
    bitis: int | None = None

    @property
    def zamanli(self) -> bool:
        return (
            self.baslangic is not None
            and self.bitis is not None
            and self.bitis > self.baslangic
        )


def dakikaya(t: time | None) -> int | None:
    """Saat -> gün başından beri dakika. Boş saat None kalır."""
    return None if t is None else t.hour * 60 + t.minute


def gruplar(araliklar: Sequence[Aralik], olcut: str) -> list[list[int]]:
    """Aynı ana denk gelen ders saatleri.

    Dönen her grup, bir kaynağın (şube/öğretmen) en fazla birine
    yerleşebileceği ders saatlerinin sıra numaralarıdır. Her ders saati en az
    bir grupta bulunur; tek başına olması da bir gruptur (aynı satıra iki ders
    konamaz kuralı oradan gelir).

    `ders_saati` ölçütünde her satır kendi grubudur.
    """
    if olcut != SAAT:
        return [[i] for i in range(len(araliklar))]

    zamanli: dict[int, list[int]] = {}
    sonuc: list[list[int]] = []
    for i, a in enumerate(araliklar):
        if a.zamanli:
            zamanli.setdefault(a.gun_index, []).append(i)
        else:
            # Saati bilinmeyen satır yalnızca kendisiyle çakışır.
            sonuc.append([i])

    # Aralık çizgesinde klikler, aralıklardan birinin başladığı noktalarda
    # oluşur: iki aralık kesişiyorsa geç başlayanın başlangıcı ikisinin de
    # içindedir. Yani bu tarama tüm çakışan çiftleri kapsar.
    for uyeler in zamanli.values():
        gorulen: set[frozenset[int]] = set()
        for nokta in sorted({araliklar[i].baslangic for i in uyeler}):
            klik = frozenset(
                i for i in uyeler
                if araliklar[i].baslangic <= nokta < araliklar[i].bitis  # type: ignore[operator]
            )
            if klik:
                gorulen.add(klik)
        # Her başlangıç noktası bir klik verir ama hepsi en büyük değildir:
        # yalnız başlayan bir aralık, biraz sonra başkasıyla kesişiyorsa iki kez
        # görünür. Kapsanan klikleri atmak modeli gereksiz kısıttan kurtarır.
        for klik in gorulen:
            if not any(klik < buyuk for buyuk in gorulen):
                sonuc.append(sorted(klik))
    # Sıra kümelerden gelir; okunur ve tekrarlanabilir olsun diye sabitlenir.
    sonuc.sort()
    return sonuc


def ortusenler(
    kimlikler: Sequence[int], araliklar: Sequence[Aralik], olcut: str
) -> dict[int, set[int]]:
    """kimlik -> onunla aynı ana denk gelen kimlikler (kendisi dahil)."""
    harita: dict[int, set[int]] = {k: {k} for k in kimlikler}
    for grup in gruplar(araliklar, olcut):
        kume = {kimlikler[i] for i in grup}
        for i in grup:
            harita[kimlikler[i]] |= kume
    return harita

File: frontend/src/lib/cakisma.ts
/** Çakışma ölçütü ve zaman ızgarasının saat denetimi.
 *
 *  Kurum, iki yerleşimin "aynı anda" sayılması için neye bakılacağını seçer:
 *  ızgaranın satırına mı, gerçek saat aralığına mı. Kural sunucuda
 *  `app/cakisma.py` içinde; buradaki karşılığı yalnızca seçimi anlatmak ve
 *  girilen saatleri denetlemek içindir.
 */
import type { CakismaOlcutu } from "./types";

export const OLCUT_SECENEKLERI: {
  id: CakismaOlcutu;
  etiket: string;
  ozet: string;
  aciklama: string;
}[] = [
  {
    id: "ders_saati",
    etiket: "Ders saati",
    ozet: "ızgaranın satırına göre",
    aciklama:
      "Salı 3. ders yalnızca Salı 3. dersle çakışır; girilen saatlere bakılmaz. Tek ve düzenli bir ızgarası olan okullar için doğru olan budur — satır zaten saatin kendisidir.",
  },
  {
    id: "saat",
    etiket: "Saat aralığı",
    ozet: "gerçek başlangıç ve bitişe göre",
    aciklama:
      "09:00–09:40 ile 09:20–10:00 ayrı satırlar olsa da çakışır. Saatleri üst üste binebilen ızgaralarda gerekir: vardiyalı okullar ya da bölümlere göre değişen ders süreleri. Saati girilmemiş satır yalnızca kendisiyle çakışır.",
  },
];

export function olcutEtiketi(olcut: CakismaOlcutu): string {
  return OLCUT_SECENEKLERI.find((s) => s.id === olcut)?.etiket ?? olcut;
}

/** "09:30" ya da "09:30:00" -> gün başından beri dakika. */
export function dakikaya(saat: string | null | undefined): number | null {
  if (!saat) return null;
  const [s, d] = saat.split(":");
  const sayi = Number(s) * 60 + Number(d);
  return Number.isFinite(sayi) ? sayi : null;
}

export type SaatSorunu = {
  tur: "cakisma" | "sira" | "eksik";
  metin: string;
};

type DenetlenenSaat = {
  name: string;
  start_time: string | null;
  end_time: string | null;
  is_break: boolean;
};

/** Bir günün saatlerindeki tutarsızlıklar.
 *
 *  Üç şeye bakar:
 *    * çakışma — iki ders saatinin aralığı üst üste biniyor,
 *    * sıra — satır sırası ile saat sırası ters,
 *    * eksik — biri girilmiş öbürü boş ya da bitiş başlangıçtan önce.
 *
 *  Teneffüsler çakışma denetiminin dışındadır: onlara ders konmuyor.
 */
export function saatSorunlari(saatler: DenetlenenSaat[]): SaatSorunu[] {
  const sorunlar: SaatSorunu[] = [];
  const araliklar = saatler.map((s) => ({
    ad: s.name || "adsız satır",
    bas: dakikaya(s.start_time),
    bit: dakikaya(s.end_time),
    ders: !s.is_break,
  }));

  for (const a of araliklar) {
    if (a.bas === null && a.bit === null) continue;
    if (a.bas === null || a.bit === null) {
      sorunlar.push({ tur: "eksik", metin: `${a.ad}: saatin bir ucu boş.` });
    } else if (a.bit <= a.bas) {
      sorunlar.push({ tur: "eksik", metin: `${a.ad}: bitiş, başlangıçtan sonra değil.` });
    }
  }

  // Sıra: satır sırası ilerlerken saat geri gitmemeli.
  const sirali = araliklar.filter((a) => a.bas !== null);
  for (let i = 1; i < sirali.length; i++) {
    if (sirali[i].bas! < sirali[i - 1].bas!) {
      sorunlar.push({
        tur: "sira",
        metin: `${sirali[i].ad}, kendinden önceki ${sirali[i - 1].ad} satırından daha erken başlıyor.`,
      });
    }
  }

  const gecerli = araliklar.filter(
    (a) => a.ders && a.bas !== null && a.bit !== null && a.bit > a.bas,
  );
  for (let i = 0; i < gecerli.length; i++) {
    for (let j = i + 1; j < gecerli.length; j++) {
      if (gecerli[i].bas! < gecerli[j].bit! && gecerli[j].bas! < gecerli[i].bit!) {
        sorunlar.push({
          tur: "cakisma",
          metin: `${gecerli[i].ad} ile ${gecerli[j].ad} saatleri üst üste biniyor.`,
        });
      }
    }
  }

  return sorunlar;
}
